internal/state/dataplane/configuration.go

```diff
@@ -471,29 +503,118 @@ func getPath(path *v1beta1.HTTPPathMatch) string {
 	return *path.Value
 }
 
+func buildMaps(servers []VirtualServer) []Map {
+	return buildAddHeaderMaps(servers)
+}
+
+func buildAddHeaderMaps(servers []VirtualServer) []Map {
+	// TODO(ciarams87): Figure out the capacity of this slice in advance
+	var maps []Map
+
+	for _, s := range servers {
+		for _, pr := range s.PathRules {
+			for _, mr := range pr.MatchRules {
+				if mr.Filters.RequestHeaderModifiers != nil {
+					for _, addHeader := range mr.Filters.RequestHeaderModifiers.Add {
+						maps = append(maps, createAddHeadersMap(addHeader.Name))
+					}
+				}
+			}
+		}
+	}
+	return deduplicateMaps(maps)
+}
+
+const (
+	// In order to prepend any passed client header values to values specified in the add headers field of request
+	// header modifiers, we need to create a map parameter regex for any string value
+	anyStringFmt = `~.*`
+)
+
+func createAddHeadersMap(name string) Map {
+	underscoreName := strings.ToLower(strings.ReplaceAll(name, "-", "_"))
+	httpVarSource := "${http_" + underscoreName + "}"
+	// FIXME(ciarams87) This var needs to be the same as what is specified in proxy_set_headers.
+	// Is there some way of sharing the name between these maps and the proxySetHeaders?
+	// FIXME(ciarams87) Ensure the length of this var is not too long - to prevent cases like:
+	// "could not build map_hash, you should increase map_hash_bucket_size: 64"
+	mapVarName := "$" + underscoreName + "_header_var"
```

Comment:
Can we consider this as a TODO, not FIXME? (so that we fix it before merging) The max length of the header is 256 (`// +kubebuilder:validation:MaxLength=256`), let's make sure NGINX doesn't fail to reload in that case